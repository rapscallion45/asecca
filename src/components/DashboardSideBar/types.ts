/*
 * Dashboard Sidebar type definitions
 *
 * These type definitions are used in the applications Dashboard Sidebar component.
 */

/**
 * Nav Bar sidebar on close callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback IDashboardSidebarOnCloseCallback
 * @returns {void} - no return value
 */
export interface IDashboardSidebarOnCloseCallback {
  (): void;
}
