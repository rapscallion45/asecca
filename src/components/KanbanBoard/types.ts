/*
 * Kanban Board type definitions
 *
 * These type definitions are used in the applications kanban board component.
 * Only make changes to this file if the corresponding change is made in the
 * Kanban Board component(s)
 */

/**
 * Kanban Board Column On Drop callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.7
 *
 * @callback IKanbanBoardColumnOnDropCallback
 * @param {number} colIndex - column index of task
 * @param {number} prevColIndex - previous column index of task
 * @param {number} taskIndex - index of task
 * @returns {void} - no return value
 */
export interface IKanbanBoardColumnOnDropCallback {
  (colIndex: number, prevColIndex: number, taskIndex: number): void;
}
