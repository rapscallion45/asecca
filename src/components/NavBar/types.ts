/*
 * Nav Bar type definitions
 *
 * These type definitions are used in the applications Nav Bar component.
 */

/**
 * Nav Bar sidebar open callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback INavBarOnSidebarOpenCallback
 * @returns {void} - no return value
 */
export interface INavBarOnSidebarOpenCallback {
  (): void;
}
