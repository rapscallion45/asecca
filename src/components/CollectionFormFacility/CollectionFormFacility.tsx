import { FC, useState, useCallback, ReactNode } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AppState, AppDispatch } from '@/redux/store';
import {
  Card,
  CardHeader,
  CardContent,
  Box,
  Button,
  Select,
  MenuItem,
  SelectChangeEvent,
  FormControl,
  Typography,
  Backdrop,
  CircularProgress,
} from '@mui/material';
import { LoadingButton } from '@mui/lab';
import {
  ICollectionFormFacilityData,
  ICollectionFormFacilityAssetCategoryFacilitiesDataPayload,
  ICollectionFormFacilityWorkflowsDataPayload,
} from '@/lib/api/api-types';
import {
  saveByCollectionId as saveFacilityByCollectionId,
  resetFacility,
  editFacility,
  fetchWorkflows,
} from '@/redux/slices/collectionFormFacilitySlice';
import DataTable from '@/components/DataTable/DataTable';
import {
  DataTableRowCellValue,
  IDataTableColumn,
} from '@/components/DataTable/types';
import columns from './collectionFormFacilityTableColumns';

/**
 * Collection Form Facility Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.17
 *
 * @typedef ICollectionFormFacilityProps
 * @prop {string} collectionId - ID string of Collection for table data API call
 */
interface ICollectionFormFacilityProps {
  collectionId: string;
}

/**
 * Collection Form Facility
 *
 * Presents the Collection Form Facility table to the user, populated with data
 * fetched from API: /api/collection/facility/api/facility
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.17
 *
 * @component
 * @param {ICollectionFormFacilityProps} props - component props
 * @returns {FC} - collection form facility table functional component
 */
const CollectionFormFacility: FC<ICollectionFormFacilityProps> = (props) => {
  const { collectionId } = props;

  /* shorthand helper for dispatching redux actions */
  const dispatch = useDispatch<AppDispatch>();

  /* get collection form facility data held in redux state */
  const {
    data,
    loading,
    error,
    saving,
    edited,
    loadingAssetCategoryFacilities,
    assetCategoryFacilities,
    loadingWorkflows,
    workflows,
  } = useSelector((state: AppState) => state.collectionFormFacility);

  /* on first load, fetch the asset categories */
  // useEffect(() => {
  //   dispatch(fetchFacilities());
  //   dispatch(fetchWorkflows());
  // }, [dispatch]);

  /**
   * Table column list
   *
   * Local state of columns list for available facility and workflow options
   *
   * @since 0.0.17
   *
   * @constant
   */
  const [colList] = useState<Array<IDataTableColumn>>(columns);

  /* Whenever Facility data updates, update column list */
  //   useEffect(() => {
  //     setColList(
  //       columns.map((col: IDataTableColumn) => {
  //         if (col.key === 'facility')
  //           return {
  //             ...col,
  //             selectOptions: assetFacilities,
  //           };
  //         if (col.key === 'workflow')
  //           return {
  //             ...col,
  //             selectOptions: assetWorkflows,
  //           };
  //         return col;
  //       })
  //     );
  //   }, [data]);

  /**
   * Handles the saving of the table data
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.17
   *
   * @method
   */
  const handleSave = useCallback(() => {
    dispatch(
      saveFacilityByCollectionId({
        data: {
          collection: collectionId,
          rows: data?.rows,
        },
      })
    );
  }, [collectionId, data, dispatch]);

  /**
   * Handles the resetting of the table data
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.17
   *
   * @method
   */
  const handleReset = useCallback(() => {
    dispatch(resetFacility());
  }, [dispatch]);

  /**
   * Handles the update of the table data
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.17
   *
   * @method
   * @param {DataTableRowCellValue} value - value to be used in update
   * @param {string} colKey - column key to be updated
   * @param {number} rowIdx - table row index to be updated
   */
  const handleEditCellValue = useCallback(
    (value: DataTableRowCellValue, colKey: string, rowIdx: number) => {
      dispatch(editFacility({ value, colKey, rowIdx }));
    },
    [dispatch]
  );

  /**
   * Handles any required logic when determining a cell's display value
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.17
   *
   * @method
   * @param {number} rowIdx - table row index to get value from
   * @param {IDataTableColumn} column - column to get value from
   * @return {DataTableRowCellValue} - cell value, can be null or undefined
   */
  const handleGetCellValue = useCallback(
    (rowIdx: number, column: IDataTableColumn): DataTableRowCellValue =>
      data.rows[rowIdx][column.key as keyof ICollectionFormFacilityData],
    [data]
  );

  /**
   * Handles requests to view selected workflow
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.17
   *
   * @method
   * @param {number} rowIdx - table row index of the action cell
   */
  /* eslint-disable @typescript-eslint/no-unused-vars */
  const handleViewWorkflow = useCallback((rowIdx: number) => {}, []);
  /* eslint-enable @typescript-eslint/no-unused-vars */

  /**
   * Handles any required logic when a facility select dropdown is changed
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.17
   *
   * @method
   * @param {SelectChangeEvent} event - facility select dropdown change event
   * @param {IDataTableColumn} column - column to update value for
   * @param {number} rowIdx - table row index to update value for
   */
  const handleFacilityChange = useCallback(
    (event: SelectChangeEvent, colKey: string, rowIdx: number) => {
      /* update the global state */
      handleEditCellValue(event.target.value, colKey, rowIdx);

      /* if we have null value, clear workflow and do not dispatch fetch */
      if (event.target.value === '') {
        /* update the global state */
        handleEditCellValue('', 'workflow', rowIdx);
        return;
      }

      /* fetch new workflow options for this facility and asset category combo */
      dispatch(
        fetchWorkflows({
          assetCategory: data.rows[rowIdx].asset_category,
          facility: event.target.value,
        })
      );
    },
    [data.rows, handleEditCellValue, dispatch]
  );

  /**
   * Handles any required logic when a workflow select dropdown is changed
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.17
   *
   * @method
   * @param {SelectChangeEvent} event - workflow select dropdown change event
   * @param {IDataTableColumn} column - column to update value for
   * @param {number} rowIdx - table row index to update value for
   */
  const handleWorkflowChange = useCallback(
    (event: SelectChangeEvent, colKey: string, rowIdx: number) => {
      /* update the global state */
      handleEditCellValue(event.target.value, colKey, rowIdx);
    },
    [handleEditCellValue]
  );

  /**
   * Helper function for rendering form table select fields
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.17
   *
   * @method
   * @param {string} colKey - table column key of the action cell
   * @param {number} rowIdx - table row index of the action cell
   * @param {string} value - current selected value
   * @param {Array<ICollectionFormFacilityAssetCategoryFacilitiesDataPayload | ICollectionFormFacilityWorkflowsDataPayload>} options - select options present
   * @param {(SelectChangeEvent, string, number) => void} handleChange - handle select change callback
   * @returns {ReactNode} - select dropdown component
   */
  const getSelectDropdown = useCallback(
    (
      colKey: string,
      rowIdx: number,
      value: string,
      options:
        | ICollectionFormFacilityAssetCategoryFacilitiesDataPayload
        | ICollectionFormFacilityWorkflowsDataPayload,
      handleChange: (
        event: SelectChangeEvent,
        colKey: string,
        rowIdx: number
      ) => void
    ): ReactNode => (
      <FormControl variant="standard" sx={{ minWidth: 200 }}>
        <Select
          id={`${colKey}-select-input`}
          value={value}
          renderValue={(selectValue: string) =>
            /* if current value is undefined, display the unassigned text */
            selectValue === undefined || selectValue === '' ? (
              <MenuItem value="">
                {columns.find((col: IDataTableColumn) => col.key === colKey)
                  ?.unassignedText || ''}
              </MenuItem>
            ) : (
              /* we have a value, display it */
              <MenuItem value={selectValue}>{selectValue}</MenuItem>
            )
          }
          onChange={(event: SelectChangeEvent) =>
            handleChange(event, colKey, rowIdx)
          }
          color="secondary"
        >
          {/* check if this dropdown can be unassigned a value */}
          {columns.find((col: IDataTableColumn) => col.key === colKey)
            ?.allowUnassigned && <MenuItem value="">Unassigned</MenuItem>}
          {/* map passed select options */}
          {options?.map((option: any) => (
            <MenuItem
              key={`${data.rows[rowIdx].asset_category}-${option}-option`}
              value={
                typeof option === 'object' && 'name' in option
                  ? option.name
                  : option
              }
            >
              {typeof option === 'object' && 'name' in option
                ? option.name
                : option}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
    ),
    [data.rows]
  );

  /**
   * Handles requests for the action component for specific column
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.17
   *
   * @method
   * @param {string} colKey - table column key of the action cell
   * @param {number} rowIdx - table row index of the action cell
   * @returns {ReactNode} - action component to present
   */
  const getActionComponent = useCallback(
    (colKey: string, rowIdx: number): ReactNode => {
      switch (colKey) {
        case 'facility':
          return getSelectDropdown(
            colKey,
            rowIdx,
            data.rows[rowIdx].facility || '',
            assetCategoryFacilities[rowIdx],
            handleFacilityChange
          );
        case 'workflow':
          return data.rows[rowIdx].facility ? (
            getSelectDropdown(
              colKey,
              rowIdx,
              data.rows[rowIdx].workflow || '',
              workflows[rowIdx],
              handleWorkflowChange
            )
          ) : (
            <Typography>Please select a facility</Typography>
          );
        case 'view_workflow':
          return data.rows[rowIdx].workflow ? (
            <Button
              color="secondary"
              variant="contained"
              onClick={() => handleViewWorkflow(rowIdx)}
              disabled={saving || loading}
              sx={{ fontSize: 'small' }}
            >
              View Workflow
            </Button>
          ) : (
            <Box sx={{ minWidth: '140.5px' }} />
          );
        default:
          return null;
      }
    },
    [
      assetCategoryFacilities,
      data.rows,
      loading,
      saving,
      workflows,
      getSelectDropdown,
      handleFacilityChange,
      handleWorkflowChange,
      handleViewWorkflow,
    ]
  );

  return (
    <Card>
      <CardHeader title="Facility" />
      <CardContent sx={{ pt: 0, position: 'relative' }}>
        <Backdrop
          open={loadingAssetCategoryFacilities || loadingWorkflows}
          sx={{ position: 'absolute', zIndex: '999' }}
        >
          <CircularProgress color="inherit" />
        </Backdrop>
        <DataTable
          name="collection form facility"
          /* filter table columns by current facility type */
          columns={colList}
          /* can edit all cells in facility table */
          editableColLabels={colList.map((col: IDataTableColumn) => col.label)}
          /* build table row props from facility data */
          rows={data.rows?.map((facility: ICollectionFormFacilityData) => ({
            label: facility.asset_category,
          }))}
          isLoading={loading}
          error={error}
          editCellValueCallback={handleEditCellValue}
          getCellValueCallback={handleGetCellValue}
          getActionComponent={getActionComponent}
        />
        <Box
          sx={{
            marginTop: 2,
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'left',
          }}
        >
          <Button
            color="secondary"
            variant="outlined"
            onClick={handleReset}
            disabled={saving || loading || !edited}
            sx={{ backgroundColor: 'common.white' }}
          >
            Reset
          </Button>
          <LoadingButton
            color="secondary"
            variant="contained"
            onClick={handleSave}
            disabled={saving || loading || !edited}
            loading={saving}
            sx={{ ml: 2 }}
          >
            Save
          </LoadingButton>
        </Box>
      </CardContent>
    </Card>
  );
};

export default CollectionFormFacility;
