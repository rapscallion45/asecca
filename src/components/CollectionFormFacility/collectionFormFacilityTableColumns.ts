import { IDataTableColumn } from '@/components/DataTable/types';

/**
 * Collection Form Facility table column defintions
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.17
 * @memberof CollectionFormFacility
 *
 * @constant
 * @type {Array<IDataTableColumn>}
 */
const collectionFormFacilityTableColumns: Array<IDataTableColumn> = [
  {
    label: 'Asset Category',
    key: 'asset_category',
    type: 'string',
  },
  {
    label: 'Facility',
    key: 'facility',
    type: 'select',
    selectOptions: [] /* select options retreived from API at runtime */,
    allowUnassigned: true,
    unassignedText: 'Unassigned',
  },
  {
    label: 'Workflow',
    key: 'workflow',
    type: 'select',
    selectOptions: [] /* select options retreived from API at runtime */,
  },
  {
    label: '',
    key: 'view_workflow',
    type: 'action',
  },
];

export default collectionFormFacilityTableColumns;
