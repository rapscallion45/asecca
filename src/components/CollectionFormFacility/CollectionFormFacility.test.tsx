import React from 'react';
import { Provider } from 'react-redux';
import {
  fireEvent,
  render,
  screen,
  waitFor,
  act,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import { fetchByCollectionId as fetchCollectionFormFacilityByCollectionId } from '@/redux/slices/collectionFormFacilitySlice';
import { ICollectionFormFacilityData } from '@/lib/api/api-types';
import store from '../../redux/store';
import columns from './collectionFormFacilityTableColumns';
import CollectionFormFacility from './CollectionFormFacility';
import collectionFormFacilityDataMock from '../../../__mocks__/CollectionForm/collectionFormFacilityDataMock';
import collectionFormFacilityAssetCategoryFacilitiesDataMock from '../../../__mocks__/CollectionForm/collectionFormFacilityAssetCategoryFacilitiesDataMock';
import collectionFormFacilityWorkflowsDataMock from '../../../__mocks__/CollectionForm/collectionFormFacilityWorkflowsDataMock';
import { IDataTableColumn } from '../DataTable/types';

/* default test query ID */
const query: string = '66135000001760012';

/* chosen data row index to check during tests */
const testDataRowIdx = 0;

/* test Logisitics Type index to check during tests */
const testAssetCategoryFacilitiesIdx = 1;
const testWorkflowsIdx = 1;

/* mock the loading panel component */
jest.mock(
  '../LoadingPanel/LoadingPanel',
  () => (show: boolean) => show ? <div data-testid="loading-panel" /> : null
);

/**
 * Collection Form Facility Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Collection Form Facility', () => {
  it('Renders correctly', async () => {
    /** Arrange */
    store.dispatch(
      fetchCollectionFormFacilityByCollectionId({
        collectionId: query,
      })
    );

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <CollectionFormFacility collectionId={query} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Table Callbacks', () => {
    beforeAll(() => jest.spyOn(window, 'fetch'));

    it('Should call API with updated values when value updated and save button clicked', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormFacilityByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Assert - get API to have been called */
      await waitFor(() => {
        expect(window.fetch).toHaveBeenCalledTimes(
          collectionFormFacilityDataMock.rows.length * 3 + 1
        );
      });
      expect(window.fetch).toHaveBeenCalledWith(
        `/api/collection/facility/api/facility?collection=${query}`,
        expect.objectContaining({
          method: 'GET',
          headers: { 'Content-Type': 'application/json' },
        })
      );
      collectionFormFacilityDataMock.rows.forEach(
        (mockDataItem: ICollectionFormFacilityData) => {
          expect(window.fetch).toHaveBeenCalledWith(
            `/api/collection/facility/api/facilities_for_asset_category?asset_category=${mockDataItem.asset_category}`,
            expect.objectContaining({
              method: 'GET',
              headers: { 'Content-Type': 'application/json' },
            })
          );
          if (mockDataItem.facility)
            expect(window.fetch).toHaveBeenCalledWith(
              `/api/collection/facility/api/workflow_for_facility?asset_category=${mockDataItem.asset_category}&facility=${mockDataItem.facility}`,
              expect.objectContaining({
                method: 'GET',
                headers: { 'Content-Type': 'application/json' },
              })
            );
        }
      );
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormFacility collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        columns.forEach((column: IDataTableColumn) => {
          if (column.key !== 'view_workflow')
            expect(screen.getByText(column.label)).toBeInTheDocument();
        });
      });
      /* check data has been rendered */
      collectionFormFacilityDataMock.rows.forEach(
        (mockData: ICollectionFormFacilityData) => {
          expect(screen.getByText(mockData.asset_category)).toBeInTheDocument();
          if (mockData.facility)
            expect(
              screen.getByDisplayValue(mockData.facility)
            ).toBeInTheDocument();
          if (mockData.workflow)
            expect(
              screen.getByDisplayValue(mockData.workflow)
            ).toBeInTheDocument();
        }
      );
      /** Act - click on dropdown, update value, click Save button */
      act(() => {
        fireEvent.change(
          screen.getByDisplayValue(
            collectionFormFacilityDataMock.rows[testDataRowIdx]
              .workflow as string
          ),
          {
            target: {
              value:
                collectionFormFacilityWorkflowsDataMock[testWorkflowsIdx].name,
            },
          }
        );
      });
      /** Assert - wait for input to update */
      await waitFor(() => {
        expect(
          screen.getByDisplayValue(
            collectionFormFacilityWorkflowsDataMock[testWorkflowsIdx].name
          )
        ).toBeInTheDocument();
      });
      /** Act - click Save button */
      act(() => {
        fireEvent(
          screen.getByText('Save'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });
      /** Assert - dropdown update and save APIs to have been called */
      await waitFor(() => {
        expect(window.fetch).toHaveBeenCalledTimes(11);
      });
      store
        .getState()
        .collectionFormFacility.data.rows.forEach(
          (mockDataItem: ICollectionFormFacilityData) => {
            expect(window.fetch).toHaveBeenCalledWith(
              `/api/collection/facility/api/facilities_for_asset_category?asset_category=${mockDataItem.asset_category}`,
              expect.objectContaining({
                method: 'GET',
                headers: { 'Content-Type': 'application/json' },
              })
            );
            if (mockDataItem.facility)
              expect(window.fetch).toHaveBeenCalledWith(
                `/api/collection/facility/api/workflow_for_facility?asset_category=${mockDataItem.asset_category}&facility=${mockDataItem.facility}`,
                expect.objectContaining({
                  method: 'GET',
                  headers: { 'Content-Type': 'application/json' },
                })
              );
          }
        );
      // expect(window.fetch).toHaveBeenCalledWith(
      //   '/api/collection/facility/api/facility',
      //   expect.objectContaining({
      //     method: 'POST',
      //     headers: { 'Content-Type': 'application/json' },
      //     body: JSON.stringify({
      //       collection: query,
      //       items: store.getState().collectionFormFacility.data.rows,
      //     }),
      //   })
      // );
    });

    it('Should reset table data when cancel button clicked', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormFacilityByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormFacility collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        columns.forEach((column: IDataTableColumn) => {
          if (column.key !== 'view_workflow')
            expect(screen.getByText(column.label)).toBeInTheDocument();
        });
      });
      /* check data has been rendered */
      collectionFormFacilityDataMock.rows.forEach(
        (mockData: ICollectionFormFacilityData) => {
          expect(screen.getByText(mockData.asset_category)).toBeInTheDocument();
          if (mockData.facility)
            expect(
              screen.getByDisplayValue(mockData.facility)
            ).toBeInTheDocument();
          if (mockData.workflow)
            expect(
              screen.getByDisplayValue(mockData.workflow)
            ).toBeInTheDocument();
        }
      );
      /** Act - click on dropdowns, update values */
      act(() => {
        fireEvent.change(
          screen.getByDisplayValue(
            collectionFormFacilityDataMock.rows[testDataRowIdx]
              .facility as string
          ),
          {
            target: {
              value:
                collectionFormFacilityAssetCategoryFacilitiesDataMock[
                  testAssetCategoryFacilitiesIdx
                ],
            },
          }
        );
      });
      act(() => {
        fireEvent.change(
          screen.getByDisplayValue(
            collectionFormFacilityDataMock.rows[testDataRowIdx]
              .workflow as string
          ),
          {
            target: {
              value:
                collectionFormFacilityWorkflowsDataMock[testWorkflowsIdx].name,
            },
          }
        );
      });
      /** Assert - wait for input to update */
      await waitFor(() => {
        expect(
          screen.getByDisplayValue(
            collectionFormFacilityWorkflowsDataMock[testWorkflowsIdx].name
          )
        ).toBeInTheDocument();
      });
      /** Act - click Reset button */
      act(() => {
        fireEvent(
          screen.getByText('Reset'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });
      /** Assert - table data should no longer hold updated values and be reset */
      await waitFor(() => {
        expect(
          screen.queryByDisplayValue(
            collectionFormFacilityAssetCategoryFacilitiesDataMock[
              testAssetCategoryFacilitiesIdx
            ]
          )
        ).toBeNull();
      });
      expect(
        screen.queryByDisplayValue(
          collectionFormFacilityWorkflowsDataMock[testWorkflowsIdx].name
        )
      ).toBeNull();
      expect(
        screen.getByDisplayValue(
          collectionFormFacilityDataMock.rows[testDataRowIdx].facility as string
        )
      ).toBeInTheDocument();
      expect(
        screen.getByDisplayValue(
          collectionFormFacilityDataMock.rows[testDataRowIdx].workflow as string
        )
      ).toBeInTheDocument();
    });
  });

  describe('Table Functions', () => {
    beforeAll(() => jest.spyOn(window, 'fetch'));

    it('Should show loading panel when Assigned Facility column updated and API called', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormFacilityByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormFacility collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        columns.forEach((column: IDataTableColumn) => {
          if (column.key !== 'view_workflow')
            expect(screen.getByText(column.label)).toBeInTheDocument();
        });
      });
      /* check data has been rendered */
      collectionFormFacilityDataMock.rows.forEach(
        (mockData: ICollectionFormFacilityData) => {
          expect(screen.getByText(mockData.asset_category)).toBeInTheDocument();
          if (mockData.facility)
            expect(
              screen.getByDisplayValue(mockData.facility)
            ).toBeInTheDocument();
          if (mockData.workflow)
            expect(
              screen.getByDisplayValue(mockData.workflow)
            ).toBeInTheDocument();
        }
      );
      /** Act - click on dropdowns, update values */
      act(() => {
        fireEvent.change(
          screen.getByDisplayValue(
            collectionFormFacilityDataMock.rows[testDataRowIdx]
              .facility as string
          ),
          {
            target: {
              value:
                collectionFormFacilityAssetCategoryFacilitiesDataMock[
                  testAssetCategoryFacilitiesIdx
                ],
            },
          }
        );
      });
      /** Assert - wait for loading panel */
      await waitFor(() => {
        expect(screen.getByTestId('loading-panel')).toBeInTheDocument();
      });
    });

    it('Should render "Please select a facility" text if Assigned Facility is not selected', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormFacilityByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormFacility collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        columns.forEach((column: IDataTableColumn) => {
          if (column.key !== 'view_workflow')
            expect(screen.getByText(column.label)).toBeInTheDocument();
        });
      });
      /* check data has been rendered */
      collectionFormFacilityDataMock.rows.forEach(
        (mockData: ICollectionFormFacilityData) => {
          expect(screen.getByText(mockData.asset_category)).toBeInTheDocument();
          if (mockData.facility)
            expect(
              screen.getByDisplayValue(mockData.facility)
            ).toBeInTheDocument();
          if (mockData.workflow)
            expect(
              screen.getByDisplayValue(mockData.workflow)
            ).toBeInTheDocument();
        }
      );
      /** Assert - check for correct text */
      await waitFor(() => {
        expect(
          screen.getByText('Please select a facility')
        ).toBeInTheDocument();
      });
      /** Act - click on dropdowns, update values */
      act(() => {
        fireEvent.change(screen.getByDisplayValue(''), {
          target: {
            value:
              collectionFormFacilityAssetCategoryFacilitiesDataMock[
                testAssetCategoryFacilitiesIdx
              ],
          },
        });
      });
      /** Assert - check that messaged is now not shown as we have selected a facility */
      await waitFor(() => {
        expect(screen.queryByText('Please select a facility')).toBeNull();
      });
      /** Act - click on dropdowns, update values */
      act(() => {
        fireEvent.change(
          screen.getByDisplayValue(
            store.getState().collectionFormFacility.data.rows[1]
              .facility as string
          ),
          {
            target: {
              value: '',
            },
          }
        );
      });
      /** Assert - check that messaged is now not shown as we have selected a facility */
      await waitFor(() => {
        expect(
          screen.getByText('Please select a facility')
        ).toBeInTheDocument();
      });
    });

    it('Should render "View Workflow" button only if Workflow is selected', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormFacilityByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormFacility collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        columns.forEach((column: IDataTableColumn) => {
          if (column.key !== 'view_workflow')
            expect(screen.getByText(column.label)).toBeInTheDocument();
        });
      });
      /* check data has been rendered */
      collectionFormFacilityDataMock.rows.forEach(
        (mockData: ICollectionFormFacilityData) => {
          expect(screen.getByText(mockData.asset_category)).toBeInTheDocument();
          if (mockData.facility)
            expect(
              screen.getByDisplayValue(mockData.facility)
            ).toBeInTheDocument();
          if (mockData.workflow)
            expect(
              screen.getByDisplayValue(mockData.workflow)
            ).toBeInTheDocument();
        }
      );
      /** Assert - check for correct button text */
      await waitFor(() => {
        expect(screen.getByText('View Workflow')).toBeInTheDocument();
      });
      /** Act - click on dropdowns, update values */
      act(() => {
        fireEvent.change(
          screen.getByDisplayValue(
            collectionFormFacilityDataMock.rows[testDataRowIdx]
              .workflow as string
          ),
          {
            target: {
              value: '',
            },
          }
        );
      });
      /** Assert - check that messaged is now not shown as we have selected a facility */
      await waitFor(() => {
        expect(screen.queryByText('View Workflow')).toBeNull();
      });
    });
  });

  describe('Table Columns', () => {
    it('Should render correctly for passed permission', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormFacilityByCollectionId({
            collectionId: query,
          })
        );
      });

      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormFacility collectionId={query} />
        </Provider>
      );

      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        columns.forEach((column: IDataTableColumn) => {
          if (column.key !== 'view_workflow')
            expect(screen.getByText(column.label)).toBeInTheDocument();
        });
      });
    });
  });
});
