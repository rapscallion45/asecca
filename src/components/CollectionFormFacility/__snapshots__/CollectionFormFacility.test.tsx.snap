// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection Form Facility Renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
      >
        Facility
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root css-1i63tak-MuiCardContent-root"
  >
    <div
      data-testid="loading-panel"
    />
    <div
      className="MuiBox-root css-kh5ii8"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1ex1hsz-MuiButtonBase-root-MuiButton-root"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        Reset
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1fddvdp-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
        disabled={true}
        id=":r0:"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        Save
      </button>
    </div>
  </div>
</div>
`;
